### PR TITLE
Stabilize `unsafe_op_in_unsafe_fn` lint

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
This makes it possible to override the level of the `unsafe_op_in_unsafe_fn`, as proposed in https://github.com/rust-lang/rust/issues/71668#issuecomment-729770896.

Tracking issue: #71668
r? @nikomatsakis cc @SimonSapin @RalfJung

# Stabilization report

This is a stabilization report for `#![feature(unsafe_block_in_unsafe_fn)]`.

## Summary

Currently, the body of unsafe functions is an unsafe block, i.e. you can perform unsafe operations inside.

The `unsafe_op_in_unsafe_fn` lint, stabilized here, can be used to change this behavior, so performing unsafe operations in unsafe functions requires an unsafe block.

For now, the lint is allow-by-default, which means that this PR does not change anything without overriding the lint level.

For more information, see [RFC 2585](https://github.com/rust-lang/rfcs/blob/master/text/2585-unsafe-block-in-unsafe-fn.md)

### Example

```rust
// An `unsafe fn` for demonstration purposes.
// Calling this is an unsafe operation.
unsafe fn unsf() {}

// #[allow(unsafe_op_in_unsafe_fn)] by default,
// the behavior of `unsafe fn` is unchanged
unsafe fn allowed() {
    // Here, no `unsafe` block is needed to
    // perform unsafe operations...
    unsf();
    
    // ...and any `unsafe` block is considered 
    // unused and is warned on by the compiler.
    unsafe {
        unsf();
    }
}

#[warn(unsafe_op_in_unsafe_fn)]
unsafe fn warned() {
    // Removing this `unsafe` block will
    // cause the compiler to emit a warning.
    // (Also, no "unused unsafe" warning will be emitted here.)
    unsafe {
        unsf();
    }
}

#[deny(unsafe_op_in_unsafe_fn)]
unsafe fn denied() {
    // Removing this `unsafe` block will
    // cause a compilation error.
    // (Also, no "unused unsafe" warning will be emitted here.)
    unsafe {
        unsf();
    }
}
```

### CODE REVIEW

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -273,6 +273,8 @@ declare_features! (
     /// Allows patterns with concurrent by-move and by-ref bindings.
     /// For example, you can write `Foo(a, ref b)` where `a` is by-move and `b` is by-ref.
     (accepted, move_ref_pattern, "1.48.0", Some(68354), None),
+    /// No longer treat an unsafe function as an unsafe block.
```
Comment (RalfJung) [inline compiler/rustc_feature/src/accepted.rs:276]: Can you mention here that this is about a lint that still needs to be separately enabled?

File: compiler/rustc_mir/src/transform/check_unsafety.rs [hunk 2]
```diff
@@ -564,17 +563,10 @@ fn is_enclosed(
...
+            tcx.hir().find(parent_id)
         {
-            if sig.header.unsafety == hir::Unsafety::Unsafe
-                && !tcx.features().unsafe_block_in_unsafe_fn
-            {
-                Some(("fn".to_string(), parent_id))
-            } else {
-                None
-            }
+            None
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir/src/transform/check_unsafety.rs:569]: I think this is incorrect and leading to the current CI failure

**Comment by crlf0710** [comment]
Triage: there's merge conflicts now.

**Comment by crlf0710** [comment]
Triage: there's merge conflicts now.

File: compiler/rustc_mir/src/transform/check_unsafety.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_unsafety.rs:569]: @LeSeulArtichaut did you plan to do further work on this PR?

Comment (LeSeulArtichaut) [reply compiler/rustc_mir/src/transform/check_unsafety.rs:569]: Yes, sorry, I need to get back to it...

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_unsafety.rs:569]: No worries, just wanted to know the status. :)

**Comment by crlf0710** [comment]
@LeSeulArtichaut Ping from triage. Any updates here?

**Comment by LeSeulArtichaut** [comment]
Opened #81110, marking as blocked on that

**Comment by LeSeulArtichaut** [comment]
#81110 was merged, so this is not blocked anymore

**Comment by nikomatsakis** [comment]
Nominating for @rust-lang/lang discussion. This would stabilize the lint, though it currently defaults to Allow.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

Per our discussion on [2021-02-02](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-02-02.md), we would like to move forward with stabilization for this lint. However, @LeSeulArtichaut, we would like to request a stabilization report giving a (brief) description of what the lint does, key test cases, what behavior is being stabilized, and any other interesting information that has come up since the RFC. This seems like a pretty straightforward case so I wouldn't expect it to be very long!

**Comment by nikomatsakis** [comment]
@rfcbot concern needs-stabilization-report

**Comment by scottmcm** [comment]
@rfcbot reviewed

(It's still not obvious to me whether this should ever become a hard error, but I'm definitely in favour of having the lint available.)

**Comment by LeSeulArtichaut** [comment]
@nikomatsakis Added a stabilization report to the OP. Let me know it if needs any modification.

**Comment by mbartlett21** [comment]
@LeSeulArtichaut 
The RFC link should probably be to <https://github.com/rust-lang/rfcs/blob/master/text/2585-unsafe-block-in-unsafe-fn.md>, not this PR.

**Comment by nikomatsakis** [comment]
@rfcbot resolve needs-stabilization-report

Looks good, @LeSeulArtichaut !

**Pull request merged**